### PR TITLE
docs: ADR on spend-limit constraint

### PR DESCRIPTION
**Pull request by brobro10000**
ADR on the recently introduced `spend-limit` constraint within the `clean()` function within the `subsidy_access_policy` model

**Description:**
Add a description of your changes here. 

**Jira:**
ENT-XXXX

**Merge checklist:**
- [ ] `./manage.py makemigrations` has been run
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.

**Post merge:**
- [ ] Ensure that your changes went out to the stage instance
- [ ] Deploy to prod instance

### CODE REVIEW

File: docs/decisions/0022-spend-limits-constraint.rst [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+======
+**Accepted** (June 2024)
+
+Context
+=======
+When a customer provisions a learner-credit-based policy, ``enterprise_subsidy``'s ``Subsidy`` model populates the field 
+``starting_balance`` with their initial deposit. Depending on which tool was used to provision the customer's subsidy,
+the ``starting_balance`` would initally be set as the sum of all policies ``spend_limit`` or a user defined ``starting_balance``.
+A pattern began to emerge where the the sum of all policies associated to a subsidy's ``spend_limit`` field found on the 
+``SubsidyAccessPolicy`` model would be greater then the ``total_deposits`` value. The ``total_deposits`` value represents the sum of
```
Comment (iloveagent57) [inline docs/decisions/0022-spend-limits-constraint.rst:14]: nit: `total_deposits` is really only the _positive_ committed transactions, right?

Comment (iloveagent57) [reply docs/decisions/0022-spend-limits-constraint.rst:14]: ...or really, just the sum of all committed deposits and adjustments.

File: docs/decisions/0022-spend-limits-constraint.rst [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+Having the sum of the ``spend_limit`` for policies be greater then the ``total_deposits`` value does not accurately represent how
+the ``spend_limit`` should be used.
+
+Decision
+========
+Validation was added in the ``SubsidyAccessPolicy`` model's ``clean()`` function to verify that a modified or created ``spend_limit``
+would adhere to the following constraint:
+
+* For ``active`` policy's, (policies with the same ``subsidy_uuid`` and ``enterprise_customer``)
```
Comment (iloveagent57) [inline docs/decisions/0022-spend-limits-constraint.rst:26]: nit: policy's -> policies

File: docs/decisions/0022-spend-limits-constraint.rst [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    * The sum of all policies ``spend_limit`` must not exceed the subsidy record's ``total_deposits``
+
+The purpose of adding the validation within the ``clean()`` function ensures modification to the field via the Django admin screen
+would be adhere to the constraint without modification to the original ``SubsidyAccesPolicy`` model. Modification of the
```
Comment (iloveagent57) [inline docs/decisions/0022-spend-limits-constraint.rst:30]: nit: "would be adhere" -> "adheres"

File: docs/decisions/0022-spend-limits-constraint.rst [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+``SubsidyAccessPolicy`` model would require a backfill heuristic which was deemed too complex due to the ``spend_limit``
+``null`` value corresponding to an unlimited ``spend_limit`` budget. 
+
+Additionally, the ``clean()`` function call was added to the  ``subsidy_access_policy`` serializer to ensure the constraints
+were applied to API's that modify the ``spend_limit`` field.
```
Comment (iloveagent57) [inline docs/decisions/0022-spend-limits-constraint.rst:35]: nit: phrase as "to ensure that this new constraint is enforced to any API views that allow modification of the ``spend_limit`` field"

File: docs/decisions/0022-spend-limits-constraint.rst [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+``SubsidyAccessPolicy`` model would require a backfill heuristic which was deemed too complex due to the ``spend_limit``
+``null`` value corresponding to an unlimited ``spend_limit`` budget. 
+
+Additionally, the ``clean()`` function call was added to the  ``subsidy_access_policy`` serializer to ensure the constraints
+were applied to API's that modify the ``spend_limit`` field. 
+
+
+Consequences
+============
+This puts a stricter definition on ``spend_limit`` (amount of spend allowable to spend on the policy level) not exceeding
```
Comment (iloveagent57) [inline docs/decisions/0022-spend-limits-constraint.rst:40]: nit: "amount of spend allowable to spend on the policy level" -> "amount of allowable spend per policy"
and "amount of spend given to spend of the subsidy level" -> "amount of allowable spend per subsidy"

File: docs/decisions/0022-spend-limits-constraint.rst [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+were applied to API's that modify the ``spend_limit`` field. 
+
+
+Consequences
+============
+This puts a stricter definition on ``spend_limit`` (amount of spend allowable to spend on the policy level) not exceeding 
+the ``total_deposits`` (amount of spend given to spend of the subsidy level). This ensures that a learner cannot spend more
+funds then the subsidy has available.
+
+Furthermore, additional steps would need to be taken when modifying ``spend_limit`` on multiple policies within the same subsidy
```
Comment (iloveagent57) [inline docs/decisions/0022-spend-limits-constraint.rst:44]: nit: phase as "additional steps are now needed in two scenarios: when modifying ``spend_limit`` across multiple policies; and when adding an ``adjustment`` to a subsidy. For example:"

**Review by iloveagent57 — APPROVED** [review]

File: docs/decisions/0023-spend-limits-constraint.rst [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+and when adding an ``adjustment`` to a subsidy.
+
+For example:
+
+* When modifying the ``spend_limit`` of polices that become inactive or set inactive
+    * For Subsidy A with ``total_deposits`` value of 50,000 with two policies.
+        * Policy A, an ``active`` policy with a ``spend_limit`` of 10,000
+        * Policy B, an ``active`` policy with a ``spend_limit`` of 40,000
+    * If Policy A were set to ``!active``
+        * Policy B's ``spend_limit`` can now be set to ``50,000`` (even if spend has occured on Policy A or B)
```
Comment (adamstankiewicz) [inline docs/decisions/0023-spend-limits-constraint.rst:54]: super nit: doesn't seem like other instances of numeric dollar values here (e.g., `10,000` and `40,000` above) are formatted with the backticks. could remove the backticks for consistency with other numeric values.

**Review by adamstankiewicz — APPROVED** [review]
LGTM. One minor suggestion might be to include link to this ADR [here](https://github.com/openedx/enterprise-access/tree/main/docs/decisions#architectural-decision-records) in the list of other ADRs (though it does look like the more recent ones are missing from the list).

File: docs/decisions/0023-spend-limits-constraint.rst (hunk 7 above)
Comment (brobro10000) [reply docs/decisions/0023-spend-limits-constraint.rst:54]: Updated in this PR 👍🏽 
https://github.com/openedx/enterprise-access/pull/528

**Pull request merged**